Add browserkit component documentation

yamiko-ninja opened the pull request:
Working on documentation for the browser kit. Fixes #954
Pull request #4310 was to the wrong branch so making a new one

Commit:
added introduction

Commit:
fixed a few mistakes

Commit:
syntax fix

Commit:
added in sections that I apln to fill out

Commit:
added links with info about clients and packagist link

Commit:
added links to index and map files

Commit:
make a basic request

Commit:
fixed spelling

Commit:
added a link snippit

Commit:
more outlines, fixed link, added more about creating a client

WouterJ commented:
ping @yamiko-ninja do you have time in the future to bring this PR to the next level? If not, please say so, then someone else can take it over.

yamiko-ninja commented:
@WouterJ ahh forgot about it! 

If someone wants to take over they can. otherwise I will put this on my todo list to do by the end of April. About to move in a few weeks.

WouterJ commented:
Hi @yamiko-ninja! Just wanted to let you know that there will be a [doc sprint day](http://symfony.com/blog/symfony-docs-hack-day-needs-you-on-may-23rd) on May 23rd. You can join in then to get quick feedback & other help. If you no longer have the motivation to finish this PR, we can also put it on the list for other people to take over.

Thanks for the bootstrapping!

yamiko-ninja commented:
I will put that day on my calendar and will do some more work on it in the meantime. I have time to work on this again now.

yamiko-ninja committed:
added form submissions and moved creating a client to top

yamiko-ninja committed:
added docs on histroy

javiereguiluz commented:
@WouterJ I have some concerns about this PR. @yamiko-ninja has done a great work bootstraping the documentation of the component. However, if we expect him to complete all the documentation, maybe that's too much for a single person (hence my concerns).

What do you think if we lower the expectations of this PR to consider it just the bootstraping of the component documentation. That way we can polish @yamiko-ninja's current work a bit and merge it. Then, we outline all the contents we want to add to this section and complete that task in further pull requests.

xabbuh commented:
I agree with you @javiereguiluz for two reasons:

1. Only bootstrapping a rough skeleton would make it easier to get something started.
1. We can then create a list of smaller TODOs which would make it easier for other contributors to pick a single simple task.

yamiko-ninja commented:
I agree. I have only made a few contributions here and there I would be fine with someone else taking over.

OskarStark commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,145 @@
+
+.. code-block:: php
+
+    namespace ACME;
```
should be ``AppBundle``

WouterJ replied:
No, we are in the components section here. It should not use framework things like bundles

Commit:
Adding documentation for cookies

yamiko-ninja commented:
I added in some docs on cookies. I think this is a good start and other can add more to it if they like. With that in mind I removed the [WIP] from the title.

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+
+    The BrowserKit component simulates the behavior of a web browser.
+
+The BrowserKit component allows you to make web request, click on links and submit forms. 
```
requests

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+Installation
+------------
+
+You can install the component in 2 different ways:
```
"two" instead of "2"

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+-----------
+
+.. note::
+    The component only provides an abstract client and does not provide any "default" backend for the HTTP layer.
```
There should be a blank line between the directive and its contents.

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+    use Symfony\Component\BrowserKit\Client as BaseClient;
+    use Symfony\Component\BrowserKit\Response;
+
+    class Client extends BaseClient {
```
We always put the brace on the next line (same for methods).

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+
+For a simple implementation of a browser based on an HTTP layer, have a look at Goutte_.
+
+For an implementation based on HttpKernelInterface, have a look at the Client provided by the :doc:`/components/http_kernel/introduction`.
```
You should enclose the interface name with literals (double backticks).

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+For an implementation based on HttpKernelInterface, have a look at the Client provided by the :doc:`/components/http_kernel/introduction`.
+
+
+Making Request
```
Requests

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+Making Request
+~~~~~~~~~~~~~~
+
+To make a request you use the client's request method. 
```
You can use the `method` role to link to the API docs of the `request()` method here.

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+~~~~~~~~~~~~~~
+
+To make a request you use the client's request method. 
+The first two arguments are for the HTTP method and the request url.
```
URL

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+Clicking Links
+~~~~~~~~~~~~~~
+
+Select a link with the crawler and pass it to the click method to click on the link.
```
Same here, you can also link to the API docs of the `click()` method.

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+History
+-------
+
+The client stores all your request allowing you to go back and forward in your history.
```
requests

xabbuh commented on components/browser_kit/introduction.rst:
```diff
@@ -0,0 +1,233 @@
+    // go forward to documentation page
+    $doc_crawler = $client->forward();
+
+You can restart the clients history with the restart method. This will also clear out the CookieJar.
```
[...] the client's history [...]

xabbuh commented:
This looks pretty good so far. Thanks for your work @yamiko-ninja! I have left you a few comments (I wasn't able to look into this too deeply yet). And could you please also update this so that each lines is wrapped after the first word that crosses the 72nd character?

yamiko-ninja commented:
@xabbuh thanks for all the corrections. I will make the changes over the weekend.

yamiko-ninja committed:
spelling and formating

yamiko-ninja committed:
fix title underline

javiereguiluz commented:
@yamiko-ninja you did a great work in this pull request. We just need a final push to finish it. That's what I'm trying to do in #6072. I've reused your original commits, so you'll get full credit for your work once the other pull request is merged. Thanks!
